

#' Annotate peaks' associated PAS and motifs
#'
#' @param peak.sites.file Peak sites table generated by function peak_calling
#' @param pas.reference.file Known PAS table
#' @param output.file
#' @param genome
#' @param pas.search.cut.1
#' @param pas.search.cut.2
#' @param polystretch_length
#' @param max_mismatch
#' @param motif.search.cut
#' @param invert_strand
#'
#' @return
#' @export
#'
#' @examples
#'
#'
#'
peak_annotation <- function(peak.sites.file = peak.sites.file,
                            pas.reference.file = pas.reference.file,
                            output.file,
                            genome = genome,
                            pas.search.cut.1 = 0,
                            pas.search.cut.2 = 300,
                            polystretch_length = 13,
                            max_mismatch = 1,
                            motif.search.cut = 300,
                            #pA_motif_min_position = 100,
                            #pA_motif_max_position = 180,
                            #AAA_motif_min_position = 210,
                            invert_strand = FALSE) {

  # Read in the peak sites table
  mode.pos = gene = seq = strand = gene.start = gene.end = mu = sigma = pi = from = to = width.fit = polyA_ID = cluster_ID = UTR3_hr = NULL
  peak.table <- utils::read.table(peak.sites.file, header = TRUE, sep=",", stringsAsFactors = FALSE)
  peak.table <- peak.table %>% dplyr::select(mode.pos, gene, seq, strand, gene.start, gene.end, UTR3_hr, mu, sigma, pi, from, to, width.fit, polyA_ID, cluster_ID)
  abnormal.id = which(peak.table$width.fit <= 0)
  peak.table <- peak.table[!peak.table$cluster_ID %in% peak.table$cluster_ID[abnormal.id],]
  # print(paste("Annotating ", nrow(peak.table), " peak coordinates after drop abnormals."))
  peak.table = peak.table %>% tidyr::drop_na()
  # print(paste("Annotating ", nrow(peak.table), " peak coordinates after drop na."))
  peak.table = peak.table %>% dplyr::filter(UTR3_hr == 'YES')
  # print(paste("Annotating ", nrow(peak.table), " peak coordinates after filtering UTR/CDS."))
  peak.table = peak.table[peak.table$sigma < 200 & peak.table$sigma > 10, ]
  # print(paste("Annotating ", nrow(peak.table), " peak coordinates after filtering out extreme sigma."))
  peak.table <- peak.table %>% dplyr::select(mode.pos, gene, seq, strand, gene.start, gene.end, mu, sigma, pi, from, to, width.fit, polyA_ID, cluster_ID)

  message('Number of peaks after pre-filtering: ', nrow(peak.table))

  return('done')



}
