

#' Annotate peaks' associated PAS and motifs
#'
#' @param peak.sites.file Peak sites table generated by function peak_calling
#' @param pas.reference.file Known PAS table
#' @param output.file Output directory
#' @param genome Full genome sequence
#' @param pas.search.cut.1 Distance (nt) from the peak mode to upstream boundary of the searching interval for PAS
#' @param pas.search.cut.2 Distance (nt) from the peak mode to downstream boundary of the searching interval for PAS
#' @param polystretch_length Length of consecutive A sequence
#' @param max_mismatch Maximal tolerance of mismatch
#' @param motif.search.cut Window width for searching specified motifs
#' @param invert_strand default FALSE
#'
#' @return A peak annotation table.
#' @export
#'
#' @examples
#' ls()
#'
#'
peak_annotation <- function(peak.sites.file = peak.sites.file,
                            pas.reference.file = pas.reference.file,
                            output.file,
                            genome = genome,
                            pas.search.cut.1 = 0,
                            pas.search.cut.2 = 300,
                            polystretch_length = 13,
                            max_mismatch = 1,
                            motif.search.cut = 300,
                            #pA_motif_min_position = 100,
                            #pA_motif_max_position = 180,
                            #AAA_motif_min_position = 210,
                            invert_strand = FALSE) {

  # Read in the peak sites table
  mode.pos = gene = seq = strand = gene.start = gene.end = mu = sigma = pi = from = to = width.fit = polyA_ID = cluster_ID = UTR3_hr = NULL
  peak.table <- utils::read.table(peak.sites.file, header = TRUE, sep=",", stringsAsFactors = FALSE)
  peak.table <- peak.table %>% dplyr::select(mode.pos, gene, seq, strand, gene.start, gene.end, UTR3_hr, mu, sigma, pi, from, to, width.fit, polyA_ID, cluster_ID)
  abnormal.id = which(peak.table$width.fit <= 0)
  peak.table <- peak.table[!peak.table$cluster_ID %in% peak.table$cluster_ID[abnormal.id],]
  peak.table = peak.table %>% tidyr::drop_na()
  peak.table = peak.table %>% dplyr::filter(UTR3_hr == 'YES')
  peak.table = peak.table[peak.table$sigma < 200 & peak.table$sigma > 10, ]
  peak.table <- peak.table %>% dplyr::select(mode.pos, gene, seq, strand, gene.start, gene.end, mu, sigma, pi, from, to, width.fit, polyA_ID, cluster_ID)

  message('Number of peaks after pre-filtering: ', nrow(peak.table))

  # read in pas.reference.file
  Chrom = Positio = Strand = mgi_symbol = NULL
  pas.ref.df = utils::read.csv(pas.reference.file)
  pas.ref.df = pas.ref.df %>% dplyr::select(Chrom, Position, Strand, mgi_symbol)

  # build gr for pas reference and peak regions
  pas.ref.df.gr = GenomicRanges::GRanges(paste0(pas.ref.df$Chrom,":", pas.ref.df$Position, ":", pas.ref.df$Strand))
  peak.table$lb.center = peak.table$mode.pos + ifelse(peak.table$strand == '-', -pas.search.cut.2, pas.search.cut.1)
  peak.table$ub.center = peak.table$mode.pos + ifelse(peak.table$strand == '-', -pas.search.cut.1, pas.search.cut.2)
  peak.table$lb.center = pmin(pmax(peak.table$lb.center, peak.table$gene.start), peak.table$gene.end)
  peak.table$ub.center = pmax(pmin(peak.table$ub.center, peak.table$gene.end), peak.table$gene.start)
  pk_gr = GenomicRanges::GRanges(paste0(peak.table$seq, ":", peak.table$lb.center, "-", peak.table$ub.center, ":", peak.table$strand))

  # find hits (known PAS)
  idx = pas.dist = NULL
  peak.table$n_PAS = peak.table$PAS_full = ""
  all_hits <-  GenomicAlignments::findOverlaps(pk_gr, pas.ref.df.gr, type = 'any')
  idx_to_annotate_known_PAS <- S4Vectors::queryHits(all_hits)
  found.id = unique(idx_to_annotate_known_PAS)
  hits.pas.df.0 = data.frame(pas.ref.df.gr[S4Vectors::subjectHits(all_hits)])
  hits.pas.df = data.frame(idx = idx_to_annotate_known_PAS, pas = hits.pas.df.0$start)
  hits.pas.df$pas.dist = (peak.table$mode.pos[hits.pas.df$idx] - hits.pas.df$pas)*((peak.table$strand[hits.pas.df$idx] == '-')*2 -1)
  hits.pas.df = hits.pas.df %>% dplyr::group_by(idx) %>% dplyr::summarise(nn = dplyr::n(), full_list = paste0(pas.dist, collapse = ','))
  peak.table$n_PAS[found.id] = hits.pas.df$nn
  peak.table$PAS_full[found.id] = hits.pas.df$full_list

  # find motif
  # generate gr object
  all.peaks <- peak.table$polyA_ID
  # strand = sub(".*:.*:.*-.*:(.*)", "\\1", all.peaks)
  # strand = plyr::mapvalues(x = strand, from = c("1", "-1"), to = c("+", "-"))
  # peak.remainder = sub(".*:(.*:.*-.*):.*", "\\1", all.peaks)
  # peaks.use = paste0(peak.remainder, ":", strand)
  peaks.use = sub(".*:(.*:.*-.*:.*)", "\\1", all.peaks)
  gr <- GenomicRanges::GRanges(peaks.use)

  ## add meta data
  ##GenomicRanges::mcols(gr)$orig.start = peak.table$start
  ##GenomicRanges::mcols(gr)$orig.end = peak.table$end
  ##GenomicRanges::mcols(gr)$helper = peak.table$helper
  #peak.table$crude.start = peak.table$mode.pos - motif.search.cut; peak.table$crude.end = peak.table$mode.pos + motif.search.cut
  #GenomicRanges::mcols(gr)$crude.start = pmin(pmax(peak.table$crude.start, peak.table$gene.start), peak.table$gene.end)
  #GenomicRanges::mcols(gr)$crude.end = pmin(pmax(peak.table$crude.end, peak.table$gene.start), peak.table$gene.end)
  #GenomicRanges::mcols(gr)$mode.pos = peak.table$mode.pos
  #GenomicRanges::mcols(gr)$polyA_ID = peak.table$polyA_ID


  print(peaks.use)
  print(gr)

  return('done')



}
